### PR TITLE
chore: optimize avatar names

### LINKED ISSUE #825: Optimize avatar names
We should optimize how we render avatar shape names.

**Testing with lots of avatar-bots** -> https://www.notion.so/decentraland/Performance-Sync-36542a31bd714de19eaeb0c53510e97b?p=d800d1361f1d41a0844bb385068d78d7&showMoveTo=true

1. enter **ZONE** at position=**91,151** (we deployed a stadium there)
2. open browser console
3. run `clientDebug.InstantiateBotsAtCoords(100, 91.2, 151.2, 36, 50)` to instantiate 100 bot inside the stadium. It may take a little while to load as the bots collections and wearables are being randomized and loaded.

- [x] Render all names in a single canvas instead of 1 canvas for each name
- [x] Implement a max avatar names cap to avoid rendering all names
- [x] Improve avatar names transparency threshold: currently avatar names transparency based on distance starts appearing at a very far away distance (and probably the threshold from 1 to 0 is too small)
- [x] Adjust rendering order so we can batch backgrounds and texts

**Measuring performance score to compare changes** -> https://www.notion.so/decentraland/Performance-Sync-36542a31bd714de19eaeb0c53510e97b?p=550b7d14369a425ba1d8a40b5785716a&showMoveTo=true

### PR DESCRIPTION
**Pull request by AjimenezDCL**
Fixes #825 

To optimize the Avatar Names we are going to:
- User the same canvas for each name
- Sort the pieces so we can batch the draws of the containers, the text and the voice chat icon.

Approach:
1) Create a new HUD called `AvatarNames`. 
2) Use the `DataStore` to keep a `PlayerStatus` collection.
3) `Minimap` refactoring to use the new `PlayerStatus` collection.
4) `AvatarNameHUD` implementation.

It must be tested with this Kernel: 
decentraland/explorer#2503

Test Link: https://play.decentraland.zone/branch/chore/avatar-names-optimization/index.html?renderer=urn:decentraland:off-chain:renderer-artifacts:chore/optimize-avatar-names

### CODE REVIEW

**Comment by AjimenezDCL** [comment]
# Measurements

## Standing Still
```
Before Optimization
PerformanceMeter - Data report step 2 - Processed values:
 * PERFORMANCE SCORE (0-100) -> 74,49
 * average FPS -> 29,59333
 * highest FPS -> 33
 * lowest FPS -> 25
 * 50 percentile (median) FPS -> 30
 * 95 percentile FPS -> 33
 * total hiccups (>0,05ms frames) -> 3 (0,33% of frames were hiccups)
 * total hiccups time (seconds) -> 0,2143326
 * total frames -> 900
 * total frames time (seconds) -> 30,00287
```

```
After Optimization
PerformanceMeter - Data report step 2 - Processed values:
 * PERFORMANCE SCORE (0-100) -> 75,92
 * average FPS -> 31,42437
 * highest FPS -> 36
 * lowest FPS -> 26
 * 50 percentile (median) FPS -> 32
 * 95 percentile FPS -> 35
 * total hiccups (>0,05ms frames) -> 5 (0,53% of frames were hiccups)
 * total hiccups time (seconds) -> 0,3589172
 * total frames -> 952
 * total frames time (seconds) -> 30,00278
```

## Moving around
```
Before Optimization
PerformanceMeter - Data report step 2 - Processed values:
 * PERFORMANCE SCORE (0-100) -> 73,21
 * average FPS -> 28,47691
 * highest FPS -> 33
 * lowest FPS -> 22
 * 50 percentile (median) FPS -> 29
 * 95 percentile FPS -> 32
 * total hiccups (>0,05ms frames) -> 9 (1,04% of frames were hiccups)
 * total hiccups time (seconds) -> 0,5629196
 * total frames -> 866
 * total frames time (seconds) -> 30,0013
```

```
After Optimization
PerformanceMeter - Data report step 2 - Processed values:
 * PERFORMANCE SCORE (0-100) -> 74,98
 * average FPS -> 30,23965
 * highest FPS -> 35
 * lowest FPS -> 25
 * 50 percentile (median) FPS -> 30
 * 95 percentile FPS -> 35
 * total hiccups (>0,05ms frames) -> 4 (0,44% of frames were hiccups)
 * total hiccups time (seconds) -> 0,2732468
 * total frames -> 918
 * total frames time (seconds) -> 30,0129
```

**Comment by AjimenezDCL** [comment]
# Measurements 

Benchmark in Build

```
Before Optimization
PerformanceMeter - Data report step 2 - Processed values:
 * PERFORMANCE SCORE (0-100) -> 26.64
 * average FPS -> 18.37345
 * highest FPS -> 19
 * lowest FPS -> 11
 * 50 percentile (median) FPS -> 19
 * 95 percentile FPS -> 19
 * total hiccups (>0.05ms frames) -> 437 (77.35% of frames were hiccups)
 * total hiccups time (seconds) -> 23.67953
 * total frames -> 565
 * total frames time (seconds) -> 30.0313
```

```
After Optimization
PerformanceMeter - Data report step 2 - Processed values:
 * PERFORMANCE SCORE (0-100) -> 39.05
 * average FPS -> 19.19595
 * highest FPS -> 20
 * lowest FPS -> 11
 * 50 percentile (median) FPS -> 19
 * 95 percentile FPS -> 20
 * total hiccups (>0.05ms frames) -> 319 (53.89% of frames were hiccups)
 * total hiccups time (seconds) -> 16.94389
 * total frames -> 592
 * total frames time (seconds) -> 30.02672
```

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs [hunk 1]
```diff
@@ -169,13 +177,15 @@ public override void OnPoolGet()
...
             model = new AvatarModel();
             lastAvatarPosition = null;
-            avatarName.SetName(String.Empty);
         }
 
         public override void Cleanup()
         {
             base.Cleanup();
 
+            if (playerStatus != null)
```
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs:186]: You'd want to null `playerStatus` too. Remember `AvatarShape` is poolable. If `playerState` is not nulled, it will go inside the if at line 123 for recycled avatars.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+            name.ForceMeshUpdate(); //To get the new bounds
+            UpdatePosition();
+        }
+
+        public void UpdatePosition()
+        {
+            if (playerStatus == null)
+                return;
+
+            var mainCamera = Camera.main;
```
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:57]: `Camera.main` should be accessed only once in the lifetime of this object to save performance.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            background = backgroundRect.GetComponent<Image>();
+            name = nameRect.GetComponent<TextMeshProUGUI>();
+            voiceChatCanvasGroup = voiceChatRect.GetComponent<CanvasGroup>();
+            voiceChatAnimator = voiceChatRect.GetComponent<Animator>();
+        }
+
+        public void SetVisibility(bool visible)
+        {
+            visibility = visible;
+            background?.gameObject.SetActive(visibility);
```
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:36]: We shouldn't be using null-conditional operator on `UnityEngine.Object` instances as they can raise the `missingReferenceException` in weird cases. This is because the equality operator is overriden on that type, and the operator skips the override.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+            UpdatePosition();
+        }
+
+        public void UpdatePosition()
+        {
+            if (playerStatus == null)
+                return;
+
+            var mainCamera = Camera.main;
+            Vector3 screenPoint = mainCamera == null ? Vector3.zero : mainCamera.WorldToViewportPoint(playerStatus.worldPosition + OFFSET);
```
Comment (BrianAmadori) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:58]: Remember that when many avatars are present, and avatars are on your back, names for avatars that are outside the viewport can be rendered. We should try to deactivate those renderers.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs (hunk 1 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs:186]: You just solved me a bug!

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 2 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:57]: Yep, I'm counting on it. This is still WIP.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 4 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:58]: They will be placed outside of the renderer and wont be visible. In any case I will deactivate them to gain that edge.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 3 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:36]: Oh, I will use a manual check then, thx!

Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:36]: In the end I decided to remove the null check and rely on the tests to prevent that edge case

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 2 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs]: Solved!

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs [hunk 5]
```diff
@@ -31,9 +30,10 @@ public class AvatarShape : BaseComponent
         public bool everythingIsLoaded;
 
         private Vector3? lastAvatarPosition = null;
-        private MinimapMetadata.MinimapUserInfo avatarUserInfo = new MinimapMetadata.MinimapUserInfo();
         bool initializedPosition = false;
 
+        private PlayerStatus playerStatus = null;
```
Comment (sandrade-dcl) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs:35]: In order to avoid confussions, maybe a better name for this `playerStatus` would be `avatarStatus`? As this data is managed by any avatar in the world and not only by our avatar... What do you think?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDController.cs [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+        private BaseDictionary<string, PlayerStatus> otherPlayersStatus => DataStore.i.player.otherPlayersStatus;
+
+        internal IAvatarNamesHUDView view;
+        internal readonly HashSet<string> trackingPlayers = new HashSet<string>();
+        internal readonly LinkedList<string> reservePlayers = new LinkedList<string>();
+
+        internal virtual IAvatarNamesHUDView CreateView() { return AvatarNamesHUDView.CreateView(); }
+
+        public AvatarNamesHUDController() : this(200) { }
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDController.cs:18]: Can we move this 200 to a constant? so we have more context about what it is

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs [hunk 7]
```diff
@@ -126,6 +117,35 @@ public override IEnumerator ApplyChanges(BaseModel newModel)
...
+            playerStatus.isTalking = model.talking;
+            playerStatus.worldPosition = entity.gameObject.transform.position;
+            if (isNew)
+                DataStore.i.player.otherPlayersStatus.Add(playerStatus.id, playerStatus);
+        }
+
+        private void Update()
+        {
+            if (playerStatus != null)
+                playerStatus.worldPosition = entity.gameObject.transform.position;
```
Comment (sandrade-dcl) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs:146]: Couldn't we do this position update in the `OnEntityTransformChanged()` instead of do it in each frame?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDView.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        public static IAvatarNamesHUDView CreateView()
+        {
+            AvatarNamesHUDView view = Instantiate(Resources.Load<GameObject>("AvatarNamesHUD")).GetComponent<AvatarNamesHUDView>();
+            view.gameObject.name = "_AvatarNamesHUD";
+            return view;
+        }
+
+        private void OnEnable()
+        {
+            canvasRect.GetComponent<Canvas>().sortingOrder = -2;
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDView.cs:41]: We should move this to somewhere common to the project like `Settings`, So in the future, if we need to change the sorting order in other parts of the project we have some visibility about the current canvas ordering. What do you think?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+            name.text = newPlayerStatus.name;
+            name.ForceMeshUpdate(); //To get the new bounds
+            UpdatePosition();
+        }
+
+        public void UpdatePosition()
+        {
+            if (mainCamera == null)
+                mainCamera = mainCamera = Camera.main;
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:57]: we should remove 1 mainCamera here

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+            {
+                background.gameObject.SetActive(false);
+                name.gameObject.SetActive(false);
+                voiceChatCanvasGroup.gameObject.SetActive(false);
+            }
+        }
+
+        public void DestroyUIElements()
+        {
+            Object.Destroy(background.gameObject);
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:94]: Shouldn't we use `GameObject.Destroy `instead of `Object `in unity objects?

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/Tests/AvatarNamesHUDView_VisualTests.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+
+            yield return null;
+
+            yield return VisualTestHelpers.TakeSnapshot();
+        }
+
+        protected override IEnumerator TearDown()
+        {
+            if (hudView != null)
+                Object.Destroy(hudView.gameObject);
```
Comment (Suduck) [inline unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/Tests/AvatarNamesHUDView_VisualTests.cs:59]: Same here, `GameObject `instead of `Object`?

**Review by Suduck — APPROVED** [review]
LGTM! Some suggestions were added.

I think we should somehow ( maybe each X seconds ) change the names of the avatars when we are moving and we are above the max numbers of avatar names.

Imagine that you are in a concert and we are moving on to where our friends are. So we are moving from the entrance to the other part of the concert. You will want to see the avatar names of the nears players and not the first 200 avatars names. 

Take this into account for the future!

**Review by sandrade-dcl — APPROVED** [review]
Reviewed and tested, LGTM! I only left some suggestions/doubts. Good work.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 9 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:57]: Oopsie

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs (hunk 5 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs:35]: It contains information about the player, such as name and the voice chat status along the position. I think it would be more confussing to have it named `avatarStatus`.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 10 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:94]: `Destroy` is a member of `Object`. `GameObject` has it because inherits `Objects`.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDView.cs (hunk 8 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDView.cs:41]: Not sure about that, this is an specific sorting set for the AvatarNames. I dont think any other part of the project should be aware of what sorting order does the AvatarNames canvas have.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 10 above)
Comment (Suduck) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs:94]: Yes, but Unity uses an implementation where he passes the time to be destroyed, not sure if it makes any change though

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDView.cs (hunk 8 above)
Comment (Suduck) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDView.cs:41]: As you wish!

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs (hunk 7 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Components/Avatar/AvatarShape.cs:146]: The entity transform position is not where the Avatar is rendered, we interpolate the target position to make it smoothly. We actually have to do it each frame or the AvatarName will be lagging behind. A nice "side-effect" is that the minimap is now smoother as well (since it can use this data).

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDController.cs (hunk 6 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesHUDController.cs]: Done!

**Comment by AjimenezDCL** [comment]
> LGTM! Some suggestions were added.
> 
> I think we should somehow ( maybe each X seconds ) change the names of the avatars when we are moving and we are above the max numbers of avatar names.
> 
> Imagine that you are in a concert and we are moving on to where our friends are. So we are moving from the entrance to the other part of the concert. You will want to see the avatar names of the nears players and not the first 200 avatars names.
> 
> Take this into account for the future!

@Suduck Yeah I was thinking the same but as I started benchmarking I noticed we can drastically increase the number of Avatar Names in screen (thanks the the batching of the UI parts). It's a nice suggestion but it wouldn't have much use until we can comfortably have more than 200 users on screen.

File: unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs (hunk 10 above)
Comment (AjimenezDCL) [reply unity-renderer/Assets/Scripts/MainScripts/DCL/Controllers/HUD/AvatarNamesHUD/AvatarNamesTracker.cs]: I mean, `GameObject` doesn't override `Destroy`. It's the same call
